### PR TITLE
feat: add wallet connection functionality

### PR DESCRIPTION
**Pull request by kristiehuang**
Adding wallet connection functionality to widget! This means that every prop is now optional; a new integrator can now see a fully functional swap widget via `<SwapWidget />` 😀   

Remaining PRs that need reviews:
- [x] updating tests #62 
- [x] updating docs https://github.com/Uniswap/docs/pull/407
- [ ] refactor wallet out of swap component #123 

Breaking changes:
* onConnectWallet prop changed to onConnectWalletClick
* jsonRpcEndpoint prop changed to jsonRpcUrlMap

Other API changes:
* new defaultChainId optional prop
* provider prop now optional
* new hideConnectionUI optional prop

wc

### CODE REVIEW

File: .nvmrc [hunk 1]
```diff
@@ -1 +1 @@
-lts/*
+16
```
Comment (JFrankfurt) [inline .nvmrc:1]: any reason *not* to use lts?

File: e2e/connect.test.tsx [hunk 2]
```diff
@@ -26,13 +26,13 @@ describe('connect', () => {
     expect(toolbar.textContent).toBe('Connect wallet to swap')
   })
 
-  describe('with jsonRpcEndpoint', () => {
+  describe('with jsonRpcUrlMap', () => {
     it('prompts for wallet connection in the Wallet', async () => {
-      component.rerender(<SwapWidget tokenList={tokens} jsonRpcEndpoint={hardhat.url} />)
+      act(() => component.rerender(<SwapWidget tokenList={tokens} jsonRpcUrlMap={{ 1: [hardhat.url] }} />))
```
Comment (JFrankfurt) [inline e2e/connect.test.tsx:31]: I'm not super familiar w/ these testing utils, but the [docs ](https://testing-library.com/docs/preact-testing-library/api/#act)seem to indicate the `act` wrapper should not be necessary.

File: package.json [hunk 3]
```diff
@@ -67,11 +67,11 @@
...
-    "@web3-react/core": "8.0.30-beta.0",
-    "@web3-react/eip1193": "8.0.23-beta.0",
-    "@web3-react/empty": "8.0.17-beta.0",
-    "@web3-react/types": "8.0.17-beta.0",
-    "@web3-react/url": "8.0.22-beta.0",
+    "@web3-react/core" : "8.0.35-beta.0",
+    "@web3-react/eip1193": "8.0.26-beta.0",
+    "@web3-react/empty": "8.0.20-beta.0",
+    "@web3-react/types": "8.0.20-beta.0",
+    "@web3-react/url": "8.0.25-beta.0",
```
Comment (JFrankfurt) [inline package.json:74]: @vm confirming these patch versions are the correct ones we should release with

File: package.json [hunk 4]
```diff
@@ -154,8 +156,9 @@
     "@uniswap/v3-core": "1.0.0",
     "@uniswap/v3-periphery": "^1.1.1",
     "@walletconnect/ethereum-provider": "^1.7.1",
-    "@web3-react/metamask": "8.0.24-beta.0",
-    "@web3-react/walletconnect": "8.0.31-beta.0",
+    "@web3-react/metamask": "8.0.27-beta.0",
+    "@web3-react/network": "8.0.27-beta.0",
+    "@web3-react/walletconnect": "8.0.35-beta.0",
```
Comment (JFrankfurt) [inline package.json:161]: I think these should be in `dependencies` not `devDependencies`

File: package.json (hunk 3 above)
Comment (vm) [reply package.json:74]: whatever is latest is what i'd use

File: package.json (hunk 4 above)
Comment (vm) [reply package.json:161]: yes ^

File: e2e/connect.test.tsx (hunk 2 above)
Comment (vm) [reply e2e/connect.test.tsx:31]: think it's because the `Web3Provider` causes issues, we have the same errors in interface atm

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+const ButtonContents = styled(Column)`
+  gap: 0.75em;
+  justify-items: center;
+`
+
+const StyledMainButton = styled(Button)`
+  background-color: ${({ theme }) => theme.container};
+  border-radius: ${({ theme }) => theme.borderRadius * 0.75}em;
+  height: 183px;
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:39]: not a huge deal, but generally if i see numbers not divisible by 4 in css i get slightly concerned. is this trying to align with another part of the page?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+  line-height: 20px;
+  white-space: pre-wrap;
+`
+
+const QRCodeWrapper = styled.div`
+  height: 110px;
+  width: 110px;
+  path {
+    /* Maximize contrast: transparent in light theme, otherwise hard-coded to light theme. */
+    fill: ${({ theme }) => (theme.container === lightTheme.container ? '#00000000' : lightTheme.container)};
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:65]: should be #000000? i think that might also be in the theme

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        if (stale) return
+        setQrCodeSvg(qrCodeSvg)
+      })
+    } else {
+      walletConnect.activate(defaultChainId).catch((e) => {
+        if (stale) return
+        onError?.(e)
+      })
+    }
+    return () => {
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:109]: what does returning a function in a use effect do?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+      const uri: string = payload.params[0]
+      if (uri) {
+        setQrUri(uri)
+      }
+    }
+    walletConnect.provider?.connector.on('display_uri', uriListener)
+
+    return () => {
+      walletConnect.events.off(URI_AVAILABLE)
+      ;(walletConnect.provider?.connector as unknown as EventEmitter | undefined)?.off('display_uri', uriListener)
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:141]: i find this line kinda hard to follow, is there a way to simplify this by naming intermediate variables?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    }
+  })
+
+  return (
+    <StyledMainButton onClick={onClick}>
+      <StyledMainButtonRow>
+        <ButtonContents>
+          <img src={logoSrc} alt={walletName} width={32} />
+          <ThemedText.Subhead1>
+            <Trans>{walletName}</Trans>
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:151]: i dont think we need to translate the wallet name

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        <QRCodeWrapper dangerouslySetInnerHTML={{ __html: qrCodeSvg }} />
+      </StyledMainButtonRow>
+    </StyledMainButton>
+  )
+}
+
+function MetaMaskButton({ walletName, logoSrc, onClick }: ButtonProps) {
+  return (
+    <StyledSmallButton onClick={onClick}>
+      <ButtonContents>
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:166]: this buttoncontents stuff can be a component since it's also used in walletconnectbutton

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        <ThemedText.Subhead1>
+          <Trans>{walletName}</Trans>
+        </ThemedText.Subhead1>
+      </ButtonContents>
+    </StyledSmallButton>
+  )
+}
+
+function NoWalletButton() {
+  const helpCenterUrl = 'https://help.uniswap.org/en/articles/5391585-how-to-get-a-wallet'
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:177]: this is constant so should live outside the component as `HELP_CENTER_URL`

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+  return (
+    <StyledSmallButton onClick={() => window.open(helpCenterUrl)}>
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:179]: this ideally should be an `<a>` styled as a button

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+      <StyledNoWalletText>
+        <Trans>I don&apos;t have a wallet</Trans>
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:181]: usually easier to do `{'I don't have a wallet'}`

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+      </StyledNoWalletText>
+    </StyledSmallButton>
+  )
+}
+
+export function ConnectWalletDialog() {
+  let defaultConnections: Web3Connection[]
+  const [firstConnector] = connections[0]
+  if (firstConnector instanceof EIP1193 || firstConnector instanceof Url) {
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:190]: name a variable for what's inside the parens, then you won't need a comment below

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+    <StyledSmallButton onClick={() => window.open(helpCenterUrl)}>
+      <StyledNoWalletText>
+        <Trans>I don&apos;t have a wallet</Trans>
+      </StyledNoWalletText>
+    </StyledSmallButton>
+  )
+}
+
+export function ConnectWalletDialog() {
+  let defaultConnections: Web3Connection[]
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:188]: defaultConnections as name is a bit confusing, what's the default mean?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+  if (firstConnector instanceof EIP1193 || firstConnector instanceof Url) {
+    // If first connector is the integrator-provided connector
+    defaultConnections = connections.slice(1)
+  } else {
+    defaultConnections = connections
+  }
+  const [mmConnection, wcTileConnection, wcPopupConnection] = defaultConnections
+  const defaultChainId = useAtomValue(defaultChainIdAtom)
+
+  const onError = (error: any) => error && console.error('web3 error:', error)
```
Comment (vm) [inline src/components/ConnectWallet/ConnectWalletDialog.tsx:199]: this should live outside the component as it's constant

Comment (vm) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:199]: the type should also be `Error`, it won't be undefined

File: src/components/ConnectWallet/ConnectedWalletChip.tsx [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+export default function ConnectedWalletChip({ disabled }: { disabled?: boolean }) {
+  // TODO(kristiehuang): AccountDialog UI does not yet exist
+  // const [open, setOpen] = useState(false)
+
+  // TODO: hover to see disconnect button is temporary; disconnection should live inside AccountDialog
+  const [hover, setHover] = useState(false)
+
+  const { account } = useWeb3React()
+  function disconnectWallet() {
```
Comment (vm) [inline src/components/ConnectWallet/ConnectedWalletChip.tsx:27]: can we disconnect to a specific wallet, since we know which is connected?

Comment (vm) [reply src/components/ConnectWallet/ConnectedWalletChip.tsx:27]: can probably do
```ts
const { connector } = useWeb3React()
connector.deactivate()
```

File: src/components/ConnectWallet/ConnectedWalletChip.tsx [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+            <Row gap={0.5}>
+              <WalletDisconnectIcon />
+              <Trans>Disconnect wallet</Trans>
+            </Row>
+          </ThemedText.Caption>
+        ) : (
+          <ThemedText.Subhead2>
+            <Row gap={0.5}>
+              <Identicon />
+              {account?.substring(0, 6)}...{account?.substring(account.length - 4)}
```
Comment (vm) [inline src/components/ConnectWallet/ConnectedWalletChip.tsx:57]: what's the situation where account is undefined here?

File: src/components/ConnectWallet/index.tsx [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+          })
+        : wallet.activate(defaultChainId)?.catch(() => {
+            console.log('Could not activate', getConnectorName(wallet))
+          })
+    )
+  }, [defaultChainId])
+
+  const { account, isActive } = useWeb3React()
+
+  const isConnected = isActive && Boolean(account)
```
Comment (vm) [inline src/components/ConnectWallet/index.tsx:31]: why does account need to be defined for isConnected to be true?

File: src/cosmos/Swap.fixture.tsx [hunk 20]
```diff
@@ -68,13 +75,17 @@ function Fixture() {
...
+      jsonRpcUrlMap={jsonRpcUrlMap}
+      defaultChainId={defaultChainId}
       provider={connector}
       theme={theme}
       tokenList={tokenList}
       width={width}
       routerUrl={routerUrl}
-      onConnectWallet={() => console.log('onConnectWallet')} // this handler is included as a test of functionality, but only logs
+      onConnectWalletClick={() => {
+        // e?.preventDefault()
```
Comment (vm) [inline src/cosmos/Swap.fixture.tsx:86]: rm

File: src/cosmos/useProvider.ts [hunk 21]
```diff
@@ -2,21 +2,48 @@ import { initializeConnector } from '@web3-react/core'
...
 
-import { INFURA_NETWORK_URLS } from './useJsonRpcEndpoint'
 import useOption from './useOption'
 
+const INFURA_KEY = process.env.INFURA_KEY
+if (INFURA_KEY === undefined) {
+  console.warn(`INFURA_KEY must be a defined environment variable to use jsonRpcUrlMap in the cosmos viewer`)
+}
+
+export const INFURA_NETWORK_URLS: { [chainId: number]: string[] } = INFURA_KEY
```
Comment (vm) [inline src/cosmos/useProvider.ts:15]: think we should return an error rather than warn here

File: src/hooks/connectWeb3/useWeb3React.tsx [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+export function getConnectorName(connector: Connector) {
+  if (connector instanceof MetaMask) return 'MetaMask'
+  if (connector instanceof WalletConnect) return 'WalletConnect'
+  if (connector instanceof Network) return 'Network'
+  if (connector instanceof JsonRpcConnector) return 'JsonRpcConnector'
+  if (connector instanceof EIP1193) return 'EIP1193'
+  return 'Unknown'
+}
+
+function getWalletFromProvider(onError: (error: Error) => void, provider?: JsonRpcProvider | Eip1193Provider) {
```
Comment (vm) [inline src/hooks/connectWeb3/useWeb3React.tsx:33]: getConnection?

File: src/hooks/connectWeb3/useWeb3React.tsx [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+  if (JsonRpcProvider.isProvider(provider)) {
+    return toWeb3Connection(initializeConnector((actions) => new JsonRpcConnector(actions, provider)))
+  } else if (JsonRpcProvider.isProvider((provider as any).provider)) {
+    throw new Error('Eip1193Bridge is experimental: pass your ethers Provider directly')
+  } else {
+    return toWeb3Connection(initializeConnector((actions) => new EIP1193({ actions, provider, onError })))
+  }
+}
+
+function getWalletFromWalletConnect(
```
Comment (vm) [inline src/hooks/connectWeb3/useWeb3React.tsx:44]: getConnection?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx (hunk 6 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:68]: https://polished.js.org/docs/#readablecolor

File: src/components/ConnectWallet/ConnectWalletDialog.tsx (hunk 5 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:42]: @kristiehuang can you confirm that this is as spec'd and an intentional % 4 exception from design?

File: src/components/ConnectWallet/ConnectWalletDialog.tsx (hunk 7 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx:111]: https://reactjs.org/docs/hooks-effect.html#effects-with-cleanup

File: src/components/ConnectWallet/ConnectWalletDialog.tsx (hunk 9 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx]: and `<Trans>` won't work unless its content is a string literal

File: src/components/ConnectWallet/ConnectedWalletChip.tsx [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+import { connections } from 'hooks/connectWeb3/useWeb3React'
+import { WalletDisconnect as WalletDisconnectIcon } from 'icons'
+import Identicon from 'icons/identicon'
+import { useState } from 'react'
+import styled from 'styled-components/macro'
+import { ThemedText } from 'theme'
+
+const AccountButton = styled(TextButton)<{ hidden?: boolean }>`
+  filter: none;
+  visibility: ${({ hidden }) => hidden && 'hidden'};
```
Comment (JFrankfurt) [inline src/components/ConnectWallet/ConnectedWalletChip.tsx:13]: add a `visible` or `unset` fallback here so it doesn't result in invalid css when hidden isn't set.

File: src/components/ConnectWallet/ConnectedWalletChip.tsx [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+`
+
+export default function ConnectedWalletChip({ disabled }: { disabled?: boolean }) {
+  // TODO(kristiehuang): AccountDialog UI does not yet exist
+  // const [open, setOpen] = useState(false)
+
+  // TODO: hover to see disconnect button is temporary; disconnection should live inside AccountDialog
```
Comment (JFrankfurt) [inline src/components/ConnectWallet/ConnectedWalletChip.tsx:20]: should resolve these todo's or remove them if they are no longer necessary

File: src/components/ConnectWallet/index.tsx (hunk 19 above)
Comment (kristiehuang) [reply src/components/ConnectWallet/index.tsx:31]: will rename as `isAccountConnected` -- if we're connected to an account, then we want to display the connected wallet chip:
Screen Shot 2022-08-02 at 12 14 48 PM

else, display this:
image

File: src/components/ConnectWallet/ConnectedWalletChip.tsx (hunk 18 above)
Comment (kristiehuang) [reply src/components/ConnectWallet/ConnectedWalletChip.tsx:57]: ConnectedWalletChip won't be rendered if account is undefined - see ConnectWallet/index.tsx L32

File: src/components/ConnectWallet/ConnectedWalletChip.tsx [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+              {account?.substring(0, 6)}...{account?.substring(account.length - 4)}
+            </Row>
+          </ThemedText.Subhead2>
+        )}
+      </AccountButton>
+      {/* {open && (
+        <Dialog color="module" onClose={() => setOpen(false)}>
+          <AccountDialog />
+        </Dialog>
+      )} */}
```
Comment (JFrankfurt) [inline src/components/ConnectWallet/ConnectedWalletChip.tsx:55]: either use, or remove this if not needed

File: src/components/ConnectWallet/index.tsx [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+import ConnectedWalletChip from './ConnectedWalletChip'
+import ConnectWallet from './ConnectWallet'
+
+interface WalletProps {
+  disabled?: boolean
+  onConnectWalletClick?: (e?: React.MouseEvent<HTMLButtonElement>) => void
+}
+
+export default function Wallet({ disabled, onConnectWalletClick }: WalletProps) {
+  // Attempt to connect eagerly on mount, and prompt switch networks when integrator's defaultChainId changes
```
Comment (JFrankfurt) [inline src/components/ConnectWallet/index.tsx:15]: > prompt switch networks when integrator's defaultChainId changes

Is this behavior (/should it be) documented?

File: .nvmrc (hunk 1 above)
Comment (kristiehuang) [reply .nvmrc:1]: https://github.com/Uniswap/widgets/pull/32/files#r898373397

File: src/components/Swap/index.tsx [hunk 28]
```diff
@@ -40,32 +42,37 @@ function getTransactionFromMap(
...
   const pendingTxs = usePendingTransactions()
   const displayTx = getTransactionFromMap(pendingTxs, displayTxHash)
 
   const onSupportedNetwork = useOnSupportedNetwork()
-  const isDisabled = !(active && onSupportedNetwork)
+  const isDisabled = !(isActive && onSupportedNetwork)
 
   const focused = useHasFocus(wrapper)
 
+  const hideConnectionUI = false // TODO(kristiehuang): add new prop to allow integrator to hide entire connection UI
```
Comment (JFrankfurt) [inline src/components/Swap/index.tsx:70]: resolve this todo

File: src/components/Widget.tsx [hunk 29]
```diff
@@ -112,6 +117,27 @@ export default function Widget(props: PropsWithChildren<WidgetProps>) {
...
+    }
+    return props.defaultChainId
+  }, [props.defaultChainId])
+  const jsonRpcUrlMap: string | JsonRpcProvider | { [chainId: number]: string[] } = useMemo(() => {
+    if (!props.jsonRpcUrlMap) return JSON_RPC_FALLBACK_ENDPOINTS
+    for (const supportedChainId of ALL_SUPPORTED_CHAIN_IDS) {
+      if (!Object.keys(props.jsonRpcUrlMap).includes(`${supportedChainId}`)) {
+        const fallbackRpc = JSON_RPC_FALLBACK_ENDPOINTS[supportedChainId as number]
+        console.warn(
+          `Did not provide a jsonRpcUrlMap for chainId: ${supportedChainId}. Falling back to free public RPC endpoint ${fallbackRpc}.`
```
Comment (JFrankfurt) [inline src/components/Widget.tsx:134]: should add a note that the free/public rpc is severely rate limited

File: src/constants/chains.ts [hunk 30]
```diff
@@ -18,18 +18,32 @@ export enum SupportedChainId {
...
+  KOVAN = 'kovan',
+  OPTIMISM = 'optimism-mainnet',
+  OPTIMISTIC_KOVAN = 'optimism-kovan',
+  ARBITRUM_ONE = 'arbitrum-mainnet',
+  ARBITRUM_RINKEBY = 'arbitrum-rinkeby',
+  POLYGON = 'polygon-mainnet',
+  POLYGON_MUMBAI = 'polygon-mumbai',
+}
+
+export const CHAIN_NAMES_TO_IDS: { [ChainName: string]: SupportedChainId } = {
```
Comment (JFrankfurt) [inline src/constants/chains.ts:35]: confirming this typing is intentional, and you didn't mean to set it to `{ [chain: ChainName]: SupportedChainId }`

File: src/cosmos/Swap.fixture.tsx (hunk 20 above)
Comment (JFrankfurt) [reply src/cosmos/Swap.fixture.tsx]: I think leaving this in as an example is fine, but I'd add a comment about the expected behavior if `e.preventDefault` *is* called.

File: src/hooks/usePoll.ts [hunk 31]
```diff
@@ -60,7 +60,7 @@ export default function usePoll<T>(
     }
 
     async function poll(ttl = Date.now() + pollingInterval) {
-      timeout = setTimeout(poll, pollingInterval) // queue the next poll
+      timeout = window.setTimeout(poll, pollingInterval) // queue the next poll
```
Comment (JFrankfurt) [inline src/hooks/usePoll.ts:63]: Why the `window.` addition?

**Review by JFrankfurt — CHANGES_REQUESTED** [review]

File: src/components/ConnectWallet/ConnectedWalletChip.tsx (hunk 18 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectedWalletChip.tsx:46]: Yeah @vm this is one of those weird situations w/ hooks where you know the component render is conditional on a hook value, but the value is accessed via that hook in the child instead of passed as a prop.

File: .nvmrc (hunk 1 above)
Comment (JFrankfurt) [reply .nvmrc:1]: Haha, I thought I remembered that! https://nodejs.org/en/about/releases/

at this point, it probably makes sense to bump it to 18 if we're going to make a change here

File: src/hooks/usePoll.ts (hunk 31 above)
Comment (kristiehuang) [reply src/hooks/usePoll.ts:63]: without explicitly setting window., `setTimeout` returns a Timeout object instead of the number we want

https://stackoverflow.com/questions/55550096/ts2322-type-timeout-is-not-assignable-to-type-number-when-running-unit-te

File: src/components/ConnectWallet/ConnectWalletDialog.tsx (hunk 13 above)
Comment (JFrankfurt) [reply src/components/ConnectWallet/ConnectWalletDialog.tsx]: This would break translations though :(

File: src/hooks/usePoll.ts (hunk 31 above)
Comment (JFrankfurt) [reply src/hooks/usePoll.ts:63]: ohhh I remember this. I think it might just be a TS issue in your editor?

using `window.` is fine!

File: src/components/ConnectWallet/index.tsx [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+            console.log('Could not activate', getConnectorName(wallet))
+          })
+    )
+  }, [defaultChainId])
+
+  const { account, isActive } = useWeb3React()
+
+  const isAccountConnected = isActive && Boolean(account)
+  return isAccountConnected ? (
+    <ConnectedWalletChip disabled={disabled} />
```
Comment (ianlapham) [inline src/components/ConnectWallet/index.tsx:33]: could we pass the account to the `ConnectedWalletChip` chip component? (so we dont have to use the hook with conditionals inside of it?)

File: src/components/Swap/index.tsx [hunk 33]
```diff
@@ -40,32 +42,36 @@ function getTransactionFromMap(
   return
 }
 
+// SwapProps also currently includes props needed for wallet connection, since the wallet connection component exists within the Swap component
+// TODO(kristiehuang): refactor WalletConnection outside of Swap component
```
Comment (ianlapham) [inline src/components/Swap/index.tsx:46]: Is this TODO still relevant?

File: src/components/Widget.tsx [hunk 34]
```diff
@@ -88,7 +92,8 @@ export type WidgetProps = {
   theme?: Theme
   locale?: SupportedLocale
   provider?: Eip1193Provider | JsonRpcProvider
-  jsonRpcEndpoint?: string | JsonRpcProvider
+  jsonRpcUrlMap?: { [chainId: number]: string[] }
+  defaultChainId?: number
```
Comment (ianlapham) [inline src/components/Widget.tsx:96]: should this be `ChainId`

**Review by ianlapham — COMMENTED** [review]
Great work, havent gone though everything but just some small comments so far

File: src/state/swap.ts [hunk 35]
```diff
@@ -27,3 +27,6 @@ export const swapAtom = atomWithImmer<Swap>({
 export const displayTxHashAtom = atom<string | undefined>(undefined)
 
 export const feeOptionsAtom = atom<FeeOptions | undefined>(undefined)
+
+// If set, allows integrator to add behavior when 'Connect wallet to swap' button is clicked
+export const onConnectWalletClickAtom = atom<(() => void | Promise<boolean>) | undefined>(undefined)
```
Comment (kristiehuang) [inline src/state/swap.ts:32]: not sure if this is the right file to put this atom in - but not sure where else to put it

File: src/components/Swap/index.tsx (hunk 33 above)
Comment (kristiehuang) [reply src/components/Swap/index.tsx:46]: https://github.com/Uniswap/widgets/pull/123 - let me know if this makes sense

Wanted to pull wallet connection out of the Swap component since it's separate concerns -- and also makes it slightly easier to build Buy/Donate/etc widgets in the future

File: src/state/swap.ts (hunk 35 above)
Comment (vm) [reply src/state/swap.ts:32]: i would make a `wallet.ts` for wallet state

**Review by vm — APPROVED** [review]
this looks great to me! would wait on approvals from others

**Review by JFrankfurt — APPROVED** [review]
